File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/List.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.unit.sp

/**
 * KinetixList / KinetixListItem — mirrors
 * `packages/ui/src/components/list.tsx`: a single-column list of rows
 * (leading icon/avatar, title, optional description, trailing content),
 * distinct from a `Table` (tabular data — not yet ported here). Radix has
 * no primitive here — the React version is plain `div`s — so this ports 1:1 with no
 * simplifications beyond the usual ones: `onSelect` (interactive rows) is
 * a nullable callback, same convention as [KinetixCheckbox]/
 * [KinetixRadioButton]; the row highlight ring the React version shows on
 * hover/focus isn't carried over (Material3's own ripple covers the
 * interaction cue, same call made for [KinetixButton]). Each row draws
 * its own bottom divider rather than a shared `divide-y` — Compose has no
 * single-property equivalent for "border between children, not around
 * them" the way Tailwind's `divide-y` does.
 */
@Composable
fun KinetixList(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Column(modifier = modifier) {
        content()
    }
}

@Composable
fun KinetixListItem(
    title: String,
    modifier: Modifier = Modifier,
    description: String? = null,
    leading: (@Composable () -> Unit)? = null,
    trailing: (@Composable () -> Unit)? = null,
    showDivider: Boolean = true,
    enabled: Boolean = true,
    onSelect: (() -> Unit)? = null,
) {
    val colors = KinetixColorScheme.current

    Column(modifier = modifier) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .clickable(enabled = enabled && onSelect != null, onClick = { onSelect?.invoke() })
                .padding(
                    horizontal = dimensionResource(R.dimen.spacing_3),
                    vertical = dimensionResource(R.dimen.spacing_3),
                ),
            horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_3)),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.invoke()
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = title,
                    color = colors.foreground,
                    fontSize = dimensionResource(R.dimen.font_size_body_md).value.sp,
                    maxLines = 1,
                )
                if (description != null) {
                    Text(
                        text = description,
                        color = colors.mutedForeground,
                        fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                        maxLines = 1,
                    )
                }
            }
            if (trailing != null) {
                Row(horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_2))) {
                    trailing()
                }
            }
        }
        if (showDivider) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(dimensionResource(R.dimen.border_width_default))
                    .background(colors.border),
            )
        }
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/List.swift
//
// List.swift — KinetixList / KinetixListItem.
//
// Mirrors packages/ui/src/components/list.tsx: a single-column list of
// rows (leading slot, title, optional description, trailing slot),
// distinct from a table. Each row draws its own bottom divider (SwiftUI
// has no `divide-y` equivalent). `onSelect` (interactive row) is a
// nullable callback, same convention as KinetixCheckbox. The hover/focus
// highlight ring isn't ported — no hover on touch.
//

import SwiftUI

public struct KinetixList<Content: View>: View {
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        VStack(spacing: 0) { content }
    }
}

public struct KinetixListItem<Leading: View, Trailing: View>: View {
    @Environment(\.kinetixColors) private var colors
    @Environment(\.isEnabled) private var isEnabled

    private let title: String
    private let description: String?
    private let onSelect: (() -> Void)?
    private let leading: Leading
    private let trailing: Trailing

    public init(
        title: String,
        description: String? = nil,
        onSelect: (() -> Void)? = nil,
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.title = title
        self.description = description
        self.onSelect = onSelect
        self.leading = leading()
        self.trailing = trailing()
    }

    private var row: some View {
        HStack(spacing: 12) {
            leading
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.kinetixBody)
                    .foregroundStyle(colors.foreground)
                    .lineLimit(1)
                if let description {
                    Text(description)
                        .font(.system(size: 13))
                        .foregroundStyle(colors.mutedForeground)
                        .lineLimit(1)
                }
            }
            Spacer(minLength: 0)
            trailing
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .contentShape(Rectangle())
        .overlay(alignment: .bottom) {
            Rectangle().fill(colors.border).frame(height: 1)
        }
    }

    public var body: some View {
        if let onSelect {
            Button(action: onSelect) { row }
                .buttonStyle(.plain)
                .opacity(isEnabled ? 1 : 0.5)
        } else {
            row
        }
    }
}

public extension KinetixListItem where Trailing == EmptyView {
    /// Leading slot, no trailing slot.
    init(
        title: String,
        description: String? = nil,
        onSelect: (() -> Void)? = nil,
        @ViewBuilder leading: () -> Leading
    ) {
        self.init(title: title, description: description, onSelect: onSelect, leading: leading) { EmptyView() }
    }
}

public extension KinetixListItem where Leading == EmptyView, Trailing == EmptyView {
    /// Title / description only.
    init(
        title: String,
        description: String? = nil,
        onSelect: (() -> Void)? = nil
    ) {
        self.init(title: title, description: description, onSelect: onSelect) { EmptyView() } trailing: { EmptyView() }
    }
}
